### ISSUE TITLE
Review the implementation of WebSocketContext::read_message

### ISSUE THREAD

**Issue by najamelan** [snapview/tungstenite-rs#76]
Following somewhat from the [discussion here](https://github.com/snapview/tungstenite-rs/pull/72#issuecomment-530927677), but I opened a separate issue, because I really wonder whether this does what it's author intended.

```rust
/// Read a message from the provided stream, if possible.
///
/// This function sends pong and close responses automatically.
/// However, it never blocks on write.
pub fn read_message<Stream>(&mut self, stream: &mut Stream) -> Result<Message>
where
    Stream: Read + Write,
{
    // Do not read from already closed connections.
    self.state.check_active()?;
    loop {
        // Since we may get ping or close, we need to reply to the messages even during read.
        // Thus we call write_pending() but ignore its blocking.
        self.write_pending(stream).no_block()?;
        // If we get here, either write blocks or we have nothing to write.
        // Thus if read blocks, just let it return WouldBlock.
        if let Some(message) = self.read_message_frame(stream)? {
            trace!("Received message {}", message);
            return Ok(message);
        }
    }
}
```
The loop seems like an iffy control structure here. After looking through the methods it calls, if I'm not mistaking, it can only loop if there is a fragmented message being read. However the comment suggests this will deal with responding to ping and close. 

So this will not respond to ping and close directly, but only on the next read. That's inconvenient for client code, especially when using through tokio-tungstenite, which means not being able to call `read_message` directly when a ping or close is returned here.

It also means that `write_pending` is called in a loop when a framed message is being read. That doesn't seem very useful. I would think in the least the call to `write pending` should be before the loop, if we really want to only respond to close/ping from the last read.

Nicer would be to respond to close/ping directly. It's not so clear to me what the best solution is. Something like this comes to mind:
```rust
// This loops so we continue reading frames when a fragmented message is coming in.
// There are 4 conditions that break from this loop:
// - errors
// - ConnectionClosed
// - nothing to read anymore (WouldBlock)
// - a complete message was read
loop {
    if let Some(message) = self.read_message_frame(stream)? {
        trace!("Received message {}", message);
        match message
        {
            // Since we may get ping or close, we need to reply to the messages even during read.
            // Thus we call write_pending() but ignore its blocking.
            Close(..) | Ping(..) => self.write_pending(stream).no_block()?,
            _ => {}
        };
        return Ok(message);
    }
}
```

The problem here is that `write_pending` might encounter an error. In that case, we have taken the message out the underlying stream, but we will never return it. It just get's lost. We just return the error.

There are several possible solutions to this problem, like storing the error on self and returning that on the next read. I'm not sure what would be best.

**Comment by application-developer-DA** [comment]
I did not quite get, what exactly does not work as expected.

For user the `read_message()` is just function to be called when the user wants to read a message. Inside the implementation this function also automatically "schedules" that the replies to the `ping`s are properly done, i.e. it ensures that we can guarantee some invariants here.

> So this will not respond to ping and close directly, but only on the next read. That's inconvenient for client code, especially when using through tokio-tungstenite, which means not being able to call read_message directly when a ping or close is returned here.

What do you mean? The client code which uses `tungstenite-rs` is either blocking (and then it works "as expected") or non-blocking (if it's built with `mio` or `tokio`). `tokio-tungstenite` relies on `tungstenite-rs` and uses `read_message()` inside `Stream`'s implementation. You can call `write_message()` after getting `Message::Ping` or `Message::Close`.

**Comment by najamelan** [comment]
> I did not quite get what exactly does not work as expected.

- First of all, the comments on the code don't really correspond to what it actually does, and figuring out what it actually does requires digging in all of the methods called here.
- write_pending will be called in a loop when a fragmented message is being read. A fragmented message is not a control message, so it seems wasteful to call `write_pending` in a loop for no reason. The only case where this would do something useful is if the send queue needs several calls before being empty and if somehow in the small delay of reading a fragment, the underlying stream get's ready to accept more data, but really I think this is very unlikely to be useful.
- For client code this means that the behavior is surprising. The docs for `read_message` say: _This function sends pong and close responses automatically._, which isn't really true. Tungstenite will respond when you next call a method that triggers `write_pending`. In the mean time I have found a [quite satisfactory solution](https://github.com/najamelan/ws_stream_tungstenite/blob/master/src/tung_websocket.rs#L106) to this, but at first, based on the documentation I just returned `None` from my stream, introducing a bug in my code. Only after analyzing and testing this did I know how to handle it. So I think for client code it isn't that much of an issue after all, but users have to know it and take it into account.

**Comment by application-developer-DA** [comment]
> First of all, the comments on the code don't really correspond to what it actually does, and figuring out what it actually does requires digging in all of the methods called here.

Hm... I think it does what it's intended to. The docs tell that it reads a message (true) and that it automatically sends pong / close responses (which is also true; had not it been true, the correct "read-only client" could not have been written).

> write_pending will be called in a loop when a fragmented message is being read. A fragmented message is not a control message, so it seems wasteful to call write_pending in a loop for no reason. The only case where this would do something useful is if the send queue needs several calls before being empty and if somehow in the small delay of reading a fragment, the underlying stream get's ready to accept more data, but really I think this is very unlikely to be useful.

`write_pending()` is required is if we don't call it, we can't ensure that we're doing some progress on replying to these pings and close messages. If we remove `write_pending()` completely, we will put a bargain of sticking to the websocket protocol on the user.

> For client code this means that the behavior is surprising. The docs for read_message say: This function sends pong and close responses automatically., which isn't really true. 

Why is the behavior surprising? It does send pong and close responses automatically. Of course we can't guarantee that it will be sent till the end after one single call to `read_message()` in non-blocking / async environment, but it does not have to be an issue (well, at least for the current version of tokio it's absolutely ok as far as I can tell).

> In the mean time I have found a quite satisfactory solution to this, but at first, based on the documentation I just returned None from my stream, introducing a bug in my code. 

I'm not quite sure what's going on in that code (I also don't have much experience with tokio-alpha), but it seems like your `poll_next()` function does approximately the same thing the `Stream` implementation in `tokio-tungstenite` does (i.e. it just reads the messages and returns them to the user).

**Comment by najamelan** [comment]
Sorry for the confusion. Maybe I don't express it well.

> `// Since we may get ping or close, we need to reply to the messages even during read.`

Since this is written in future tense and because this is in a loop, it makes it appear as if this will respond to a potential control frame received in the next statement in the loop. It doesn't. Combined with the docs of `read_message` which make it sound like when this method returns a close frame it will _automatically_ respond to it, which it doesn't, future calls to any method triggering `write_pending` will eventually, it made me wonder if the actual behavior is what the original author of the code intended, because the comments and docs don't correlate with the code. I would suggest:

```
// Since we may have received a ping or close during the last read operation, in which case 
// a response is waiting in the send queue, we try to flush the queue here to make sure that get's 
// sent out eventually even if the client code only calls `read_message`.
```

The loop actually does nothing for responding to control frames. It only allows for fragmented data frames to be read because `read_message_frame` only reads one frame at a time, so it doesn't make sense to have the call to `write_pending` inside the loop, it just wastes cpu cycles.

> Why is the behavior surprising?

Because the docs give the impression that upon receiving the CloseFrame, the response is already sent _automatically_, that is, without needing further action from the client code. In order to have the response sent out, further calls to `write_pending` are needed until `write_pending` no longer returns `WouldBlock`. That's not automatic. Note that further calls to `read_message` might also do the trick, but there's no way of telling, since the `WouldBlock` is being swallowed here.

> I'm not quite sure what's going on in that code

It turns the websocket into a stream of `Vec<u8>` to implement `AsyncRead`/`AsyncWrite` on top of that. So I can't return the close frame to the user (only bytes in a binary message should be returned). At first I didn't know how to handle that situation, since I can only return None or an error here. To make sure the extra read is done in order to trigger the sending of the response, I call `poll_next` recursively when a close frame is returned from tokio-tungstenite. That way for client code it is indeed _automatic_ as long as the `write_pending` doesn't return `WouldBlock`, since I can't really handle that situation with the interface of tokio-tungstenite.

**Comment by application-developer-DA** [comment]
> Since this is written in future tense and because this is in a loop, it makes it appear as if this will respond to a potential control frame received in the next statement in the loop. It doesn't. Combined with the docs of read_message which make it sound like when this method returns a close frame it will automatically respond to it, which it doesn't, future calls to any method triggering write_pending will eventually

Ok, I think I got your point. Based on docs you had a feeling that once `read_message()` returns, the user can be sure that the response to ping / close has already been sent, right? Which in case of working with non-blocking sockets won't be true (the response in this case will be "scheduled" and eventually sent).

I think it should be clear for a user who uses `tungstenite-rs` with non-blocking sockets, that the message won't be "fully" automatically sent and that it'll be sent eventually as it's clear that if we wanted to ensure that the response is really sent, we would have to block the function execution on `WouldBlock` and continue writing until the response is sent. For blocking sockets the behavior would be more or less expected.

While this is true, I must admit, that I agree that docs do not reflect well enough the way the current implementation is written as I've just found some cases when even this invariant won't be met: if you receive a ping / close, the `read_message_frame()` / `do_close()` will add the corresponding answer to the sent queue, but since the `read_message_frame()` returns this message back, we'll also quit the `loop` and return the message to the user. In the previous version of `tungstenite-rs` it was different -- we did not return close message and the rest of the "control" to the user in `enum`, so in this case we would made another iteration in the `loop` until we get `WouldBlock` or a frame we can return to the user. Back then the documentation was perfectly correct, but now... it's not quite precise, so I do agree with you on that. Thanks for bringing this 👍And this is actually one of the reasons we have a `loop` there and a `write_pending()` inside the loop and not outside of it.

To summarise, if I got you right, you had 2 concerns:
1. The documentation is imprecise. Agree. We can update it.
2. `write_pending()` should not be part of the `loop`, it can be placed outside of the loop. At the moment I would leave it like that since it's still required in some cases due to the fact that `read_one_frame()` may return `None`. Like for instance if we're in the process of closing, in this case calling `write_pending()` in the loop ensures that after getting `None` on `read_one_frame()`, we enter the part of the `write_pending()` which verifies the `closing_state` and checks the role of the socket and returns `ConnectionClosed` (the part we discussed recently in a different issue).

**Comment by najamelan** [comment]
Ah, I feel we finally understand each other. 

As far as documentation I have filed a PR which tries to improve the docs, so that can be used as a starting point if you want.  

>  At the moment I would leave it like that since it's still required in some cases due to the fact that read_one_frame() may return None. 

I might be missing something, but I have the impression the only case in which `read_message_frame` returns `None` is when a fragmented data frame is being read. It will never return `None` and thus not loop if any information relevant to closing was found. Eg. it will not return `None` when reading a ping, reading a closeframe or when the underlying connection is closed. Thus in all of those cases it won't loop.

I think strictly the only case in which the subsequent call to `write_pending` actually does anything useful is if the first one was blocked on backpressure and thus has not entirely emptied the queue, and somehow while reading a frame from a fragmented data message, the underlying connection now can take some more data. I think it's a very rare edge case which is coincidential (eg. there is no causal effect between the reading and the backpressure here), that it seems this is just wasting CPU cycles for no good reason in practice.

Since there is no hard guarantee here that calling `read_message` will empty the entire send queue, so there is really no obligation to do these extra calls to `write_pending`, just in case the connection could take some more data by accident. It doesn't change anything semantically.

**Comment by application-developer-DA** [comment]
I think this place of code has been reviewed while we were discussing this issue and other related items from recent PRs, so I believe this issue can be closed now since the questions have been clarified (the PR with documentation improvements has been merged and placing `write_pending()` outside of the loop can be done once we touch this place while working on other features).

**Comment by najamelan** [comment]
> placing write_pending() outside of the loop can be done once we touch this place while working on other features.

I'm not quite sure what this means. It sounds like it would just be forgotten. I can file a PR to fix this. If you just don't agree that this should be fixed, that's fine too. And since this is in essence just a performance issue, if you want the change to be backed up by benchmarks, I can write some benchmarks for that.

**Comment by application-developer-DA** [comment]
Well, from what I can see, having this function inside the loop indeed does not bring any benefit, so I hope we won't introduce any regression by this change, however it's not obvious that `read_message_frame()` returns `None` only in case of fragmented messages (perhaps we should have some test(s) to cover this case).

I'm not sure, how big the performance benefit will be in case of moving `write_pending()` outside of the loop.

**Comment by najamelan** [comment]
Ok, I will double check the code and try to find time to write a benchmark.
